webxrlayers-1.bs

```diff
@@ -0,0 +1,730 @@
+                                                   optional XRProjectionLayerInit init);
+  Promise&lt;XRQuadLayer&gt; requestQuadLayer(GLenum textureTarget, 
+                                       XRLayerInit init);
+  Promise&lt;XRCylinderLayer&gt; requestCylinderLayer(GLenum textureTarget, 
+                                               XRLayerInit init);
+  Promise&lt;XREquirectLayer&gt; requestEquirectLayer(GLenum textureTarget, 
+                                               XRLayerInit init);
+  Promise&lt;XRCubeLayer&gt; requestCubeLayer(XRLayerInit init);
+
+  XRWebGLSubImage? getSubImage(XRLayer layer); // for mono layers
+  XRWebGLSubImage? getViewSubImage(XRLayer layer, XRView view); // for stereo layers
+};
+</pre>
+
+Each {{XRWebGLLayerFactory}} has a <dfn for="XRWebGLLayerFactory">context</dfn> object which is an instance 
+of either a {{WebGLRenderingContext}} or a {{WebGL2RenderingContext}}.
+
+Each {{XRWebGLLayerFactory}} has an associated <dfn for="XRWebGLLayerFactory">session</dfn>, which is the 
+{{XRSession}} it was created with.
+
+Each {{XRLayer}} created through {{XRWebGLLayerFactory}} has an internal <dfn for="XRLayer">colorTextures</dfn> which is an |array| of {{WebGLTexture}} for color textures and 
+an internal <dfn for="XRLayer">depthStencilTextures</dfn> whcih is an |array| of {{WebGLTexture}}</dfn>for depth/stencil textures.
+
+<div class="algorithm" data-algorithm="construct-webgl-layer">
+
+The <dfn constructor for="XRWebGLLayerFactory">XRWebGLLayerFactory(|session|, |context|)</dfn> constructor 
+MUST perform the following steps when invoked:
+
+  1. Let |factory| be a new {{XRWebGLLayerFactory}}
+  1. If |session|'s [=ended=] value is <code>true</code>, throw an {{InvalidStateError}} and abort these steps.
+  1. If |context| is lost, throw an {{InvalidStateError}} and abort these steps.
+  1. If |session| is not an [=immersive session=] or |context|'s [=XR compatible=] boolean is <code>false</code>, throw an {{InvalidStateError}} and abort these steps.
+  1. Initialize |factory|'s [=XRWebGLLayer/context=] to |context|.
+  1. Initialize |factory|'s [=XRWebGLLayer/session=] to |session|.
+  1. Return |factory|.
+
+</div>
+
+ISSUE: can a user construct more than 1 factory?
+
+The {{nativeProjectionScaleFactor}} function returns the value that the |session|'s [=recommended WebGL framebuffer resolution=] 
+must be multiplied by to yield the |session|'s [=native WebGL framebuffer resolution=].
+
+<div class="algorithm" data-algorithm="requestProjectionLayerAlgo">
+The <dfn method for="XRWebGLLayerFactory">requestProjectionLayer(GLenum textureTarget, optional XRProjectionLayerInit init)</dfn> method creates a new {{XRProjectionLayer}}.
+
+When this method is invoked, it MUST run the following steps:
+
+  1. Let |promise| be [=a new Promise=].
+  1. If |session|'s [=ended=] value is <code>true</code>, throw an {{InvalidStateError}} and abort these steps.
```

Comment:
For any of the failure cases in an algorithm that returns a promise, the logic should follow this pattern. Changes needs to be made throughout this algorithm and the others.

```suggestion
  1. If |session|'s [=ended=] value is <code>true</code>, [=reject=] |promise| with an {{InvalidStateError}} and return it.
```